### PR TITLE
Support display of timezone for datetime and datetime64

### LINKED ISSUE #835: Display timezone information for datetime
**Describe what enhancement you'd like to have**

For now, when we select a timestamp, it doesn't tell you what timezone it is used. Internally if users don't specify timezone for datetime/datetime64 column, we use UTC timezone by default, but sometimes, users will specify timezone for a datetime column, without displaying timezone, it could be confusing.

```
SELECT _tp_time FROM stream;
```

More details https://github.com/timeplus-io/proton/discussions/833. Thanks @candlerb to bring up this issue.

### PR DESCRIPTION
**Pull request by Jasmine-ge**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:
close #835

### CODE REVIEW

File: tests/stream/test_stream_smoke/0098_fixed_issues2.yaml [hunk 1]
```diff
@@ -340,3 +340,73 @@ tests:
...
+            query_type: table
+            depends_on_stream: fixed_issues2_stream
+            wait: 2
+            query: insert into fixed_issues2_stream (t1, t2, t3) values (to_time('6/9/2024 13:28','Europe/London'), to_time('6/9/2024 13:28','Europe/London'), to_time('6/9/2024 13:28','Europe/London'));
+
+          - client: python
+            query_id: fixed-issues2-1
+            query_type: table
+            wait: 2
+            query: select t1, t2, t3 from table(fixed_issues2_stream);
```
Comment (yl-lisen) [inline tests/stream/test_stream_smoke/0098_fixed_issues2.yaml]: prefer to add functional test instead of smoke test in https://github.com/timeplus-io/proton/tree/c01bc3f18122361607b442dd4c7daa9fae3c5117/tests/queries_ported/0_stateless

File: src/IO/WriteHelpers.h [hunk 2]
```diff
@@ -794,6 +807,20 @@ inline void writeDateTimeText(DateTime64 datetime64, UInt32 scale, WriteBuffer &
     }
 }
 
+
+/// In the format YYYY-MM-DD HH:MM:SS.NNNNNNNNN+time zone, according to the specified time zone.
+template <char date_delimeter = '-', char time_delimeter = ':', char between_date_time_delimiter = ' '>
+inline void writeDateTimeTextWIthZone(DateTime64 datetime64, UInt32 scale, WriteBuffer & buf, const DateLUTImpl & time_zone = DateLUT::instance())
```
Comment (yl-lisen) [inline src/IO/WriteHelpers.h]: `writeDateTimeTextWithTimeZone`

File: src/IO/WriteHelpers.h [hunk 3]
```diff
@@ -760,6 +760,19 @@ inline void writeDateTimeText(time_t datetime, WriteBuffer & buf, const DateLUTI
     writeDateTimeText<date_delimeter, time_delimeter, between_date_time_delimiter>(LocalDateTime(datetime, time_zone), buf);
 }
 
+/// In the format YYYY-MM-DD HH:MM:SS+time zone, according to the specified time zone.
+template <char date_delimeter = '-', char time_delimeter = ':', char between_date_time_delimiter = ' '>
+inline void writeDateTimeTextWIthZone(time_t datetime, WriteBuffer & buf, const DateLUTImpl & time_zone = DateLUT::instance())
+{
+    writeDateTimeText<date_delimeter, time_delimeter, between_date_time_delimiter>(LocalDateTime(datetime, time_zone), buf);
```
Comment (yl-lisen) [inline src/IO/WriteHelpers.h]:
```suggestion
    writeDateTimeText<date_delimeter, time_delimeter, between_date_time_delimiter>(datetime, buf,  time_zone);
```

File: src/IO/WriteHelpers.h [hunk 4]
```diff
@@ -760,6 +760,19 @@ inline void writeDateTimeText(time_t datetime, WriteBuffer & buf, const DateLUTI
...
+    int hour = time_zone.timezoneOffset(datetime) / 3600;
+    buf.write(hour >= 0 ? '+' : '-');
+    int abs_hour = std::abs(hour);
+    buf.write(&digits100[abs_hour * 2], 2);
+    buf.write(':');
+    buf.write(&digits100[0], 2);
```
Comment (yl-lisen) [inline src/IO/WriteHelpers.h:773]: `buf.write(':00');`

File: src/IO/WriteHelpers.h [hunk 5]
```diff
@@ -760,6 +760,19 @@ inline void writeDateTimeText(time_t datetime, WriteBuffer & buf, const DateLUTI
     writeDateTimeText<date_delimeter, time_delimeter, between_date_time_delimiter>(LocalDateTime(datetime, time_zone), buf);
 }
 
+/// In the format YYYY-MM-DD HH:MM:SS+time zone, according to the specified time zone.
+template <char date_delimeter = '-', char time_delimeter = ':', char between_date_time_delimiter = ' '>
+inline void writeDateTimeTextWIthZone(time_t datetime, WriteBuffer & buf, const DateLUTImpl & time_zone = DateLUT::instance())
```
Comment (yl-lisen) [inline src/IO/WriteHelpers.h]: `writeDateTimeTextWithTimeZone`

File: src/IO/WriteHelpers.h (hunk 4 above)
Comment (yl-lisen) [inline src/IO/WriteHelpers.h]:
```c++
auto offset = time_zone.timezoneOffset(datetime);
auto asb_offset = std::abs(offset);
auto hours = asb_offset / 3600;
auto minutes = asb_offset % 3600 / 60;
buf.write(offset >= 0 ? '+' : '-');
buf.write(&digits100[hours * 2], 2);
buf.write(':');
buf.write(&digits100[minutes * 2], 2);
```

File: src/DataTypes/Serializations/SerializationDateTime.cpp [hunk 6]
```diff
@@ -46,7 +46,10 @@ void SerializationDateTime::serializeText(const IColumn & column, size_t row_num
     switch (settings.date_time_output_format)
     {
         case FormatSettings::DateTimeOutputFormat::Simple:
-            writeDateTimeText(value, ostr, time_zone);
+            if (has_explicit_time_zone)
+                writeDateTimeTextWIthZone(value, ostr, time_zone);
```
Comment (candlerb) [inline src/DataTypes/Serializations/SerializationDateTime.cpp]: s/WIthZone/WithZone/ (and elsewhere)

File: tests/queries/0_stateless/00515_enhanced_time_zones.reference
Comment (yl-lisen) [inline tests/queries/0_stateless/00515_enhanced_time_zones.reference:1]: doesn't modify these files. just `queries_ported` is ok

File: src/IO/WriteHelpers.h [hunk 7]
```diff
@@ -765,7 +765,7 @@ template <char date_delimeter = '-', char time_delimeter = ':', char between_dat
 inline void writeDateTimeTextWithTimeZone(time_t datetime, WriteBuffer & buf, const DateLUTImpl & time_zone = DateLUT::instance())
 {
     writeDateTimeText<date_delimeter, time_delimeter, between_date_time_delimiter>(datetime, buf,  time_zone);
-    auto offset = time_zone.timezoneOffset(datetime);
+    auto offset = time_zone.timezoneOffset(std::abs(datetime));
     auto asb_offset = std::abs(offset);
     auto hours = asb_offset / 3600;
     auto minutes = asb_offset % 3600 / 60;
```
Comment (yl-lisen) [inline src/IO/WriteHelpers.h]: need to test some timezones with `negative offset`, such as `2024-01-01 00:00:00-01:00`

File: src/IO/WriteHelpers.h [hunk 8]
```diff
@@ -820,20 +820,15 @@ inline void writeDateTimeTextWithTimeZone(DateTime64 datetime64, UInt32 scale, W
...
-    auto offset = time_zone.timezoneOffset(components.whole);
+    auto offset = time_zone.timezoneOffset(std::abs(components.whole));
     auto asb_offset = std::abs(offset);
     auto hours = asb_offset / 3600;
     auto minutes = asb_offset % 3600 / 60;
     buf.write(offset >= 0 ? '+' : '-');
     buf.write(&digits100[hours * 2], 2);
     buf.write(':');
     buf.write(&digits100[minutes * 2], 2);
+
```
Comment (yl-lisen) [inline src/IO/WriteHelpers.h]: remove this empty line

File: tests/queries/0_stateless/00515_enhanced_time_zones.reference [hunk 9]
```diff
@@ -1,6 +1,6 @@
-2017-12-05 10:37:47
-2017-12-05 10:37:47
-2017-12-05 10:37:47
+2017-12-05 10:37:47+05:30
+2017-12-05 10:37:47+05:30
+2017-12-05 10:37:47+05:30
```
Comment (yl-lisen) [inline tests/queries/0_stateless/00515_enhanced_time_zones.reference]: can we parse this time format correctly from a string ?

File: src/IO/WriteHelpers.h (hunk 7 above)
Comment (Jasmine-ge) [reply src/IO/WriteHelpers.h]: There are some existed cases and the performance is correct
image

File: tests/queries/0_stateless/00515_enhanced_time_zones.reference (hunk 9 above)
Comment (Jasmine-ge) [reply tests/queries/0_stateless/00515_enhanced_time_zones.reference]: According to the test case, we can do this
image

**Comment by jovezhong** [comment]
I would suggest creating build 1.6.1 after this PR is merged, since it will change how the datetime with explicit timezone will be displayed in console and to_string function. Too many patch build for 1.5.x already and this change need a bit more attention.

File: src/IO/ReadHelpers.cpp [hunk 10]
```diff
@@ -998,31 +998,57 @@ ReturnType readDateTimeTextFallback(time_t & datetime, ReadBuffer & buf, const D
     static constexpr auto date_broken_down_length = 10;
     /// hh:mm:ss
     static constexpr auto time_broken_down_length = 8;
-    /// YYYY-MM-DD hh:mm:ss
-    static constexpr auto date_time_broken_down_length = date_broken_down_length + 1 + time_broken_down_length;
+    /// +zz:zz
+    static constexpr auto zone_broken_down_length = 6;
```
Comment (yl-lisen) [inline src/IO/ReadHelpers.cpp:1003]: scoped our changes as follows:
```c++
/// proton: starts
... our code changes
/// proton: ends
```

Comment (yl-lisen) [reply src/IO/ReadHelpers.cpp:1003]: same for other where

**Review by yl-lisen — APPROVED** [review]
